// host-environment/cjs/isomorphic.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const merge_1 = require("./merge");
const to_json_1 = require("./to-json");
/**
 * Information about the host environment that the code is running in.
 */
exports.host = {
    global: global,
    os: getOSInfo(),
    node: getNodeInfo(),
    browser: false,
    env: process.env,
    merge: merge_1.merge,
    toJSON: to_json_1.toJSON,
};
// Merge the global `host` object (if it exists)
merge_1.mergeGlobalHost(exports.host, exports.host.global.host);
/**
 * Returns information about the current Node.js host.
 */
function getNodeInfo() {
    let versionParts = /(\d+)\.(\d+)(?:\.(\d+))?/.exec(process.version);
    let major = parseInt(versionParts[1], 10) || 0;
    let minor = parseInt(versionParts[2], 10) || 0;
    let patch = parseInt(versionParts[3], 10) || 0;
    return {
        version: parseFloat(`${major}.${minor}`),
        majorVersion: major,
        minorVersion: minor,
        patchVersion: patch,
        [`v${major}`]: true,
    };
}
/**
 * Returns information about the current operating system.
 */
function getOSInfo() {
    let windows = process.platform.startsWith("win");
    let mac = process.platform === "darwin";
    let linux = !windows && !mac;
    return { windows, mac, linux };
}
